plugins/module_utils/ansible_freeipa_module.py

```diff
@@ -566,6 +566,21 @@ def gen_intersection_list(user_list, res_list):
     return list(set(res_list or []).intersection(set(user_list or [])))
 
 
+def concat_attr_list(res, *args):
+    """
+    Get the lists of members to pass on `gen_*` methods as a res_list argument.
+
+    This function should be used to get members (usually users,
+    external users, hosts, external hosts, etc.) with any action and any state.
+
+    It is returning the concatenation of all attributes provided by user.
+    """
+    res_list = []
+    for attribute in args:
+        res_list += res.get(attribute, [])
```

Comment:
Shouldn't there be a test to make sure that the concatenation is done on same types only (lists or tuples)?

Reply:
I'm not sure it's necessary.  It's not a problem for this code if the concatenation is done on different types (list and list, tuple and tuple, or list and tuple).
Example of Mix list and tuple:
```
res = {'externalhost': ('external.host', 'internal.host'), 'memberhost_host': ['master.ipadomain.test', 'client.ipadomain.test']}
>>> args = ('memberhost_host', 'externalhost')
>>> res_list = []
>>> for attribute in args:
...     res_list += res.get(attribute, [])
...     
>>> print(res_list)
['master.ipadomain.test', 'client.ipadomain.test', 'external.host', 'internal.host']
```

Example for tuple and tuple:
```
res = {'externalhost': ('external.host', 'internal.host'), 'memberhost_host': ('master.ipadomain.test', 'client.ipadomain.test')}
>>> args = ('memberhost_host', 'externalhost')
>>> res_list = []
>>> for attribute in args:
...     res_list += res.get(attribute, [])
...     
>>> print(res_list)
['master.ipadomain.test', 'client.ipadomain.test', 'external.host', 'internal.host']
```

Reply:
The `+=` operator is different than the `+` operator (shame on Python!)

```
[rjeffman@corellia ~]$ python
Python 3.11.1 (main, Jan  6 2023, 00:00:00) [GCC 12.2.1 20221121 (Red Hat 12.2.1-4)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> a = [1,2]
>>> b = (3,4,5)
>>> c = 6
>>> d = []
>>> d += a
>>> d += b
>>> d += c
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not iterable
>>> d
[1, 2, 3, 4, 5]
```

But I believe we should still check for tuple and list. If it is a scalar, it will fail.

Reply:
@rjeffman , will something like this work?
```
>>> for attribute in args:
...     arg = res.get(attribute, [])
...     if isinstance(arg, tuple) or isinstance(arg, list):
...         res_list += res.get(attribute, [])
... 
```

Reply:
I think something like this would be easier to read:
```
for attribute in args:
    arg = res.get(attribute, [])
    if not isinstance(arg, (list, tuple)):
        arg = [arg]
    res_list += arg 
```

We already have similar code in the plugins. One day we might even want to have something like `ensure_sequence(value)`.

Reply:
BTW... `ensure_sequence(value)` is not good for "today", as some of the existing code require a list (and a string is also a sequence, so this need some more thought).